state/ports_ops.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"github.com/juju/errors"
+	jujutxn "github.com/juju/txn"
+	"gopkg.in/mgo.v2/bson"
+	"gopkg.in/mgo.v2/txn"
+)
+
+var (
+	_ ModelOperation = (*openClosePortsOperation)(nil)
+)
+
+type openClosePortsOperation struct {
+	p               *Ports
+	openPortRanges  []PortRange
+	closePortRanges []PortRange
+
+	updatedPortList []PortRange
+}
+
+// Build implements ModelOperartion.
+func (op *openClosePortsOperation) Build(attempt int) ([]txn.Op, error) {
+	annotateErr := func(err error) error {
+		return errors.Annotatef(err, "cannot open/close ports")
+	}
+	annotateOpenErr := func(err error, portRange PortRange) error {
+		return errors.Annotatef(err, "cannot open ports %v", portRange)
+	}
+	annotateCloseErr := func(err error, portRange PortRange) error {
+		return errors.Annotatef(err, "cannot close ports %v", portRange)
+	}
+
+	var createPortsDoc = op.p.areNew
+	if attempt > 0 {
+		if err := checkModelActive(op.p.st); err != nil {
+			return nil, annotateErr(errors.Trace(err))
+		}
+		if err := op.verifySubnetAliveWhenSet(); err != nil {
+			return nil, annotateErr(errors.Trace(err))
+		}
+		if err := op.p.Refresh(); err != nil {
+			if !errors.IsNotFound(err) {
+				return nil, annotateErr(errors.Trace(err))
+			}
+
+			// Ports doc not found; we need to add a new one.
+			createPortsDoc = true
+		}
+	}
+
+	ops := []txn.Op{
+		assertModelActiveOp(op.p.st.ModelUUID()),
+	}
+
+	// Start with a clean copy of the ranges from the ports document
+	var portListModified bool
+	op.updatedPortList = append(op.updatedPortList[0:0], op.p.doc.Ports...)
+
+	// Check for conflicts opening each port range and update the final port list
+	for _, openPortRange := range op.openPortRanges {
+		var alreadyExists bool
+		for _, existingPorts := range op.updatedPortList {
+			if existingPorts == openPortRange {
+				alreadyExists = true
+				break // nothing to do
+			}
+
+			if err := existingPorts.CheckConflicts(openPortRange); err != nil {
+				return nil, annotateOpenErr(errors.Trace(err), openPortRange)
+			}
+		}
+
+		if !alreadyExists {
+			op.updatedPortList = append(op.updatedPortList, openPortRange)
+			ops = append(ops, assertUnitNotDeadOp(op.p.st, openPortRange.UnitName))
+			portListModified = true
+		}
+	}
+
+	// Check for conflicts closing each port range and update the final port list
+	for _, closePortRange := range op.closePortRanges {
+		var foundAtIndex = -1
+		for i, existingPorts := range op.updatedPortList {
+			if existingPorts == closePortRange {
+				foundAtIndex = i
+				continue
+			}
+
+			if err := existingPorts.CheckConflicts(closePortRange); err != nil && existingPorts.UnitName == closePortRange.UnitName {
```

Comment:
This case would indeed trigger a conflict because of [this](https://github.com/juju/juju/blob/35b5d4f397ee30fd08b86be4298dd1c54c6abc43/state/ports.go#L131-L133) check.

However, the `ClosePort` call would still fail before the changes although the [80-100] range would remain open even though the close port call would subsequently fail. Given that we want all of these calls to be transactional, I think that treating this as a conflict is probably OK?

Ideally, we would need to make port range handling a bit more intelligent, i.e. in the above scenario, my expectation would be to split [80-100] into [80-89], [91, 100] when processing the close instead of complaining about a conflict. Obviously, this would need to be introduced in a separate PR.